optics-core/src/Optics/Internal/Optic.hs

```diff
@@ -66,10 +67,19 @@ import Unsafe.Coerce (unsafeCoerce)
 -- The parameters @s@ and @t@ represent the "big" structure,
 -- whereas @a@ and @b@ represent the "small" structure.
 --
-newtype Optic (k :: OpticKind) (is :: IxList) s t a b = Optic
-  { getOptic :: forall p i. Profunctor p
-             => Optic_ k p i (Curry is i) s t a b
-  }
+newtype Optic (k :: OpticKind) (is :: IxList) s t a b
+  = Optic (forall p i. Profunctor p => Optic_ k p i (Curry is i) s t a b)
+
+-- | Strip the newtype wrapper off.
+getOptic
+  :: Profunctor p
+  => Optic k is s t a b
+  -> Optic_ k p i (Curry is i) s t a b
+-- Note: This is not part of the definition of 'Optic' because it needs to be
+-- marked INLINE for GHC to optimize away profunctor classes when profiling.
+-- See https://github.com/well-typed/optics/issues/324 for more details.
+getOptic (Optic o) = o
+{-# INLINE getOptic #-}
```

Comment:
doesn't `{-# INLINE getOptic #-}` just work when `getOptic` (selector) function is defined with record syntax?

Reply:
I think the inline should not be necessary, since newtypes are advertised as "zero cost" wrappers. So one would expect those to be automatically inline. Afaik the newtypes are compiled via coercions. If this gets in the way of optimizations, this feels even more like something GHC could improve.

Reply:
> doesn't {-# INLINE getOptic #-} just work when getOptic (selector) function is defined with record syntax? 

It doesn't seem possible to do that then.

> I think the inline should not be necessary, since newtypes are advertised as "zero cost" wrappers.

Yes, it's incredibly bizarre. However, removing the INLINE pragma doesn't improve the situation.

I think it might be connected to this documentation bit:

> -fprof-auto
>
> All bindings not marked INLINE, whether exported or not, top level or nested, will be given automatic SCC annotations. Functions marked INLINE must be given a cost centre manually.